crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -0,0 +1,380 @@
+                signature.clone(),
+                ApplyFunction { id: variants[0], dispatches_to_multiple_functions: false },
+            );
+        }
+    }
+    apply_functions
+}
+
+fn function_id_to_field(function_id: FunctionId) -> FieldElement {
+    (function_id.to_usize() as u128).into()
+}
+
+/// Creates an apply function for the given signature and variants
+fn create_apply_function(
+    ssa: &mut Ssa,
+    signature: &FunctionSignature,
+    function_ids: &[FunctionId],
+) -> FunctionId {
+    assert!(!function_ids.is_empty());
+    ssa.add_fn(|id| {
+        let mut function_builder = FunctionBuilder::new("apply".to_string(), id, signature.runtime);
+        let target_id = function_builder.add_parameter(Type::Numeric(NumericType::NativeField));
+        let params_ids =
+            vecmap(signature.parameters.clone(), |typ| function_builder.add_parameter(typ));
+
+        let mut previous_target_block = None;
+        for (index, function_id) in function_ids.iter().enumerate() {
+            let is_last = index == function_ids.len() - 1;
+            let mut next_function_block = None;
+
+            let function_id_constant = function_builder.numeric_constant(
+                function_id_to_field(*function_id),
+                Type::Numeric(NumericType::NativeField),
+            );
+            let condition =
+                function_builder.insert_binary(target_id, BinaryOp::Eq, function_id_constant);
+
+            // If it's not the last function to dispatch, crate an if statement
+            if !is_last {
+                next_function_block = Some(function_builder.insert_block());
+                let executor_block = function_builder.insert_block();
+
+                function_builder.terminate_with_jmpif(
+                    condition,
+                    executor_block,
+                    next_function_block.unwrap(),
+                );
+                function_builder.switch_to_block(executor_block);
+            } else {
+                // Else just constrain the condition
+                function_builder.insert_constrain(condition);
+            }
+            // Find the target block or build it if necessary
+            let target_block = match previous_target_block {
+                Some(block) => {
+                    let current_block = function_builder.current_block();
+                    build_return_block(
+                        &mut function_builder,
+                        current_block,
+                        signature.returns.clone(),
+                        Some(block),
+                    )
+                }
+                None => {
+                    let current_block = function_builder.current_block();
+                    build_return_block(
+                        &mut function_builder,
+                        current_block,
+                        signature.returns.clone(),
+                        None,
+                    )
+                }
+            };
```

Comment:
```suggestion
            let current_block = function_builder.current_block();
            let target_block = build_return_block(
                &mut function_builder,
                current_block,
                signature.returns.clone(),
                previous_target_block,
            );
```

Reply:
oops, rewrote that part so much that both branches ended up the same :)